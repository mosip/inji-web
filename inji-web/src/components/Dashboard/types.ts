import { Location } from "react-router-dom";

export type DropdownItem = {
    label: string;
    onClick: () => void;
    textColor: string;
    key: string;
};

export type SidebarItemType = {
    icon: React.ReactNode;
    text: string;
    path: string;
    key: string;
};

export type SidebarItemProps = {
    icon: React.ReactNode;
    text: string;
    path: string;
    isActive: boolean;
    isCollapsed: boolean;
};

export type SearchCredentialProps = {
    issuerContainerBorderRadius?: string;
}

export type FAQPageProps = {
    backUrl?: string;
    withHome?: boolean;
};

export type CredentialTypesPageProps = {
    backUrl?: string;
};

export type InfoFieldProps = {
    label?: string;
    value?: string;
};

export type CollapseButtonProps = {
    isCollapsed: boolean;
    onClick: () => void;
    className?: string;
};

export type FooterProps = {
    footerRef: React.RefObject<HTMLDivElement>;
};

export type HeaderProps = {
    headerRef: React.RefObject<HTMLDivElement>;
};

export type DashboardHeaderProps = {
    headerRef: React.RefObject<HTMLDivElement>;
    headerHeight: number;
};

export type SideBarSvgIconProps = {
    outline: string;
    navUrl: string;
    location: Location;
}

export type User = {
    displayName: string;
    profilePictureUrl: string;
    email:string;
};

export type ErrorType = {
    errorCode: string;
    errorMessage: string;
};

export type UserContextType = {
    user: User | null;
    walletId: string | null;
    error: ErrorType | null;
    fetchUserProfile: () => Promise<{user: User; walletId: string}>;
    saveUser: (user: User) => void;
    removeUser: () => void;
    isLoading: boolean;
};
