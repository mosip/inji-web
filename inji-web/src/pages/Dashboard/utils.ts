export const convertStringIntoPascalCase = (text: string | null) => {
    if (text != null) {
        return text
            .toLocaleLowerCase()
            .split(" ")
            .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
            .join(" ");
    }
};
