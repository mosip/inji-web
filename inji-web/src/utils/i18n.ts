import i18n from "i18next";
import {initReactI18next} from "react-i18next";
import en from '../locales/en.json';
import fr from '../locales/fr.json';
import ta from '../locales/ta.json';
import hi from '../locales/hi.json';
import kn from '../locales/kn.json';
import ar from '../locales/ar.json';
import {storage} from "./storage";
import {DisplayArrayObject, LanguageObject} from "../types/data";

const resources = {en, ta, kn, hi, fr, ar};

export const LanguagesSupported: LanguageObject[] = [
    {label: "English", value: 'en'},
    {label: "தமிழ்", value: 'ta'},
    {label: "ಕನ್ನಡ", value: 'kn'},
    {label: "हिंदी", value: 'hi'},
    {label: "Français", value: 'fr'},
    {label: "عربي", value: 'ar'}
]

export const defaultLanguage = window._env_.DEFAULT_LANG;

const selected_language = storage.getItem(storage.SELECTED_LANGUAGE);
i18n
    .use(initReactI18next) // passes i18n down to react-i18next
    .init({
        resources,
        lng: selected_language ? selected_language : defaultLanguage,
        fallbackLng: defaultLanguage,
        interpolation: {
            escapeValue: false // react already safes from xss => https://www.i18next.com/translation-function/interpolation#unescape
        },
    });

export const switchLanguage = async (language: string) => {
    storage.setItem(storage.SELECTED_LANGUAGE, language);
    await i18n.changeLanguage(language);
}
export const getObjectForCurrentLanguage = (displayArray: DisplayArrayObject[], language: string = i18n.language) => {
    let resp = displayArray.filter(displayObj => (displayObj.language === language || displayObj.locale === language))[0];
    if (!resp) {
        resp = displayArray.filter(displayObj => (displayObj.language === defaultLanguage || displayObj.locale === defaultLanguage))[0];
    }
    return resp;
}

export const getDirCurrentLanguage = (language: string) => {
    return isRTL(language) ? 'rtl' : 'ltr';
}

export const isRTL = (language:string) => {
    return language === 'ar';
}
