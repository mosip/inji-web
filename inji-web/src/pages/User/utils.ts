import { User } from "../../components/Dashboard/types";
import { KEYS } from "../../utils/constants";
import {ROUTES} from "../../constants/Routes";

export const convertStringIntoPascalCase = (text: string | undefined) => {
    return (
        text?.toLocaleLowerCase()
            .split(' ')
            .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
            .join(' ')
    );
};

export const navigateToUserHome = (navigate: any) =>
    navigate(ROUTES.USER_HOME);

export const validateWalletUnlockStatus = (
    cachedWalletId: string | null,
    storageWalletId: string | null,
    navigate: (path: string) => void,
    user: User
) => {
    if (cachedWalletId && (cachedWalletId === storageWalletId)) {
        console.info('Wallet is unlocked!');
    } else {
        console.warn(
            'Wallet exists but is locked, redirecting to `/pin` to unlock the wallet.'
        );
        if(user) {
            navigate('/pin');
            localStorage.removeItem(KEYS.WALLET_ID);
        }
    }
};