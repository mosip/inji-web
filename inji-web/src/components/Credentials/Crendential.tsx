import React, {useState} from "react";
import {getObjectForCurrentLanguage} from "../../utils/i18n";
import {ItemBox} from "../Common/ItemBox";
import {generateCodeChallenge, generateRandomString} from "../../utils/misc";
import {addNewSession} from "../../utils/sessions";
import {useSelector} from "react-redux";
import {api} from "../../utils/api";
import {CredentialProps} from "../../types/components";
import {CodeChallengeObject, CredentialConfigurationObject} from "../../types/data";
import {RootState} from "../../types/redux";
import {DataShareExpiryModal} from "../../modals/DataShareExpiryModal";

export const Credential: React.FC<CredentialProps> = (props) => {
    const selectedIssuer = useSelector((state: RootState) => state.issuers);
    const [credentialExpiry, setCredentialExpiry] = useState<boolean>(false);
    const language = useSelector((state: RootState) => state.common.language);
    const filteredCredentialConfig: CredentialConfigurationObject = props.credentialWellknown.credential_configurations_supported[props.credentialId];
    const credentialObject = getObjectForCurrentLanguage(filteredCredentialConfig.display, language);
    const vcStorageExpiryLimitInTimes = useSelector((state: RootState) => state.common.vcStorageExpiryLimitInTimes);

    const onSuccess = (defaultVCStorageExpiryLimit: number = vcStorageExpiryLimitInTimes) => {
        const state = generateRandomString();
        const code_challenge: CodeChallengeObject = generateCodeChallenge(state);
        window.open(api.authorization(selectedIssuer.selected_issuer, props.credentialWellknown, filteredCredentialConfig, state, code_challenge), '_self', 'noopener');
        addNewSession({
            selectedIssuer: selectedIssuer.selected_issuer,
            certificateId: props.credentialId,
            codeVerifier: state,
            vcStorageExpiryLimitInTimes: defaultVCStorageExpiryLimit,
            state: state,
        });
    }

    return <React.Fragment>
        <ItemBox index={props.index}
                 url={credentialObject.logo.url}
                 title={credentialObject.name}
                 onClick={() => {selectedIssuer.selected_issuer.qr_code_type === 'OnlineSharing' ? setCredentialExpiry(true) : onSuccess(-1)} } />
                        { credentialExpiry &&
                            <DataShareExpiryModal onCancel={() => setCredentialExpiry(false)}
                                                  onSuccess={onSuccess}
                                                  credentialName={credentialObject.name}
                                                  credentialLogo={credentialObject.logo.url}/>
                        }
    </React.Fragment>
}

