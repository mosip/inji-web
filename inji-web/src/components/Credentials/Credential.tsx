import React, {useState} from "react";
import {getCredentialTypeDisplayObjectForCurrentLanguage,} from "../../utils/i18n";
import {ItemBox} from "../Common/ItemBox";
import {generateCodeChallenge, generateRandomString} from "../../utils/misc";
import {addNewSession} from "../../utils/sessions";
import {useSelector} from "react-redux";
import {api} from "../../utils/api";
import {CredentialProps} from "../../types/components";
import {CodeChallengeObject, CredentialConfigurationObject} from "../../types/data";

import {RootState} from "../../types/redux";
import {DataShareExpiryModal} from "../../modals/DataShareExpiryModal";
import {useUser} from "../../hooks/User/useUser";
import { buildAuthorizationUrl } from "../../utils/misc";

export const Credential: React.FC<CredentialProps> = (props) => {
    const credentials = useSelector(
        (state: RootState) => state.credentials.credentials
    );

    const authorizationEndpoint = credentials?.authorization_endpoint;
    const grantTypesSupported = credentials?.grant_types_supported;

    const selectedIssuer = useSelector(
        (state: RootState) => state.issuers.selected_issuer
    );
    const [credentialExpiry, setCredentialExpiry] = useState<boolean>(false);
    const language = useSelector((state: RootState) => state.common.language);
    const filteredCredentialConfig: CredentialConfigurationObject =
        props.credentialWellknown;
    const credentialObject = getCredentialTypeDisplayObjectForCurrentLanguage(
        filteredCredentialConfig.display,
        language
    );
    const vcStorageExpiryLimitInTimes = useSelector(
        (state: RootState) => state.common.vcStorageExpiryLimitInTimes
    );
    const {isUserLoggedIn} = useUser();

    const onSuccess = async (
        defaultVCStorageExpiryLimit: number = vcStorageExpiryLimitInTimes
    ) => {
        const state = generateRandomString();
        const code_challenge: CodeChallengeObject =
            generateCodeChallenge(state);
        addNewSession({
            selectedIssuer: selectedIssuer,
            selectedCredentialType: {type: filteredCredentialConfig.name, displayObj: filteredCredentialConfig.display},
            codeVerifier: state,
            vcStorageExpiryLimitInTimes: isNaN(defaultVCStorageExpiryLimit)
                ? vcStorageExpiryLimitInTimes
                : defaultVCStorageExpiryLimit,
            state: state
        });

        if (
            validateIfAuthServerSupportRequiredGrantTypes(grantTypesSupported)
        ) {
            const url = buildAuthorizationUrl(
                selectedIssuer,
                filteredCredentialConfig,
                state,
                code_challenge,
                authorizationEndpoint!
              );
              window.open(url, "_self", "noopener");
        } else {
            props.setErrorObj({
                code: "errors.authorizationGrantTypeNotSupportedByWallet.code",
                message:
                    "errors.authorizationGrantTypeNotSupportedByWallet.message"
            });
        }
    };

    const validateIfAuthServerSupportRequiredGrantTypes = (
        grantTypesSupported: string[] | undefined
    ) => {
        const supportedGrantTypes = ["authorization_code"];

        if (grantTypesSupported) {
            return grantTypesSupported.some((grantType: string) =>
                supportedGrantTypes.includes(grantType)
            );
        }
        return false;
    };

    return (
        <React.Fragment>
            <ItemBox
                index={props.index}
                url={credentialObject.logo}
                title={credentialObject.name}
                onClick={() => {
                    isUserLoggedIn() || selectedIssuer.qr_code_type !== "OnlineSharing"
                                                    ? onSuccess(-1)
                                                    : setCredentialExpiry(true);
                }}
            />
            {credentialExpiry && (
                <DataShareExpiryModal
                    onCancel={() => setCredentialExpiry(false)}
                    onSuccess={onSuccess}
                    credentialName={credentialObject.name}
                    credentialLogo={credentialObject.logo}
                />
            )}
        </React.Fragment>
    );
};