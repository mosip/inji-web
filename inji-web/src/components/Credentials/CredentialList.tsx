import React from "react";
import {Credential} from "./Crendential";
import {useSelector} from "react-redux";
import {CredentialWellknownObject} from "../../types/data";
import {RootState} from "../../types/redux";
import {EmptyListContainer} from "../Common/EmptyListContainer";
import {useTranslation} from "react-i18next";
import {RequestStatus} from "../../hooks/useFetch";
import {SpinningLoader} from "../Common/SpinningLoader";
import {CredentialListProps} from "../../types/components";
import {HeaderTile} from "../Common/HeaderTile";

export const CredentialList: React.FC<CredentialListProps> = ({state}) => {

    const credentials = useSelector((state: RootState) => state.credentials);
    const {t} = useTranslation("CredentialsPage");

    if (state === RequestStatus.LOADING) {
        return <SpinningLoader />
    }

    if (state === RequestStatus.ERROR || !credentials?.filtered_credentials || (credentials?.filtered_credentials && credentials?.filtered_credentials.length === 0)) {
        return <div>
                <HeaderTile content={t("containerHeading")}/>
                <EmptyListContainer content={t("emptyContainerContent")}/>
            </div>
    }

    return <React.Fragment>
            <HeaderTile content={t("containerHeading")}/>
            <div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-4">
                {credentials?.filtered_credentials && credentials?.filtered_credentials.map((credential: CredentialWellknownObject, index: number) => (
                    <Credential credential={credential} key={index} index={index}/>
                ))}
            </div>
    </React.Fragment>
}


