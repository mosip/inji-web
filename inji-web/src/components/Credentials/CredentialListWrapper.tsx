import React from "react";
import {CredentialList} from "./CredentialList";
import {RequestStatus} from "../../hooks/useFetch";

interface CredentialListWrapperProps {
    state: RequestStatus;
    className: string;
}

export const CredentialListWrapper: React.FC<CredentialListWrapperProps> = ({state, className}) => {
    return (
        <div
            data-testid="Credential-List-Container"
            className={className}
        >
            <CredentialList state={state}/>
        </div>
    );
};