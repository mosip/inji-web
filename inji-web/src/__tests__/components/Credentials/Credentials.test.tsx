import React from 'react';
import {fireEvent, render, screen} from '@testing-library/react';
import {Credential} from "../../../components/Credentials/Crendential";
import {CredentialWellknownObject, DisplayArrayObject, LogoObject} from "../../../types/data";
import {Provider} from "react-redux";
import {reduxStore} from "../../../redux/reduxStore";
import {getObjectForCurrentLanguage} from "../../../utils/i18n";

global.window._env_ = {
    DEFAULT_FAVICON: "favicon.ico",
    DEFAULT_FONT_URL: "",
    DEFAULT_THEME: "purple_theme",
    DEFAULT_TITLE: "Inji Web Test",
    DEFAULT_LANG: 'en'
};

const getCredentialObject = (): CredentialWellknownObject => {
    return {
        format: "ldp_vc",
        id: "id",
        scope: "mosip_ldp_vc",
        display: {
            name: "Name",
            language: "en",
            locale: "en",
            logo: {
                url: "https://url.com",
                alt_text: "alt text of the url"
            },
            title: "Title",
            description: "Description",
        }
    }
};

describe.skip("Test Credentials Item Layout",() => {
    test('check the presence of the container', () => {
        const clickHandler = jest.fn();
        const credential: CredentialWellknownObject = getCredentialObject();
        jest.spyOn(require('../../../utils/i18n'), 'getObjectForCurrentLanguage').mockReturnValue(credential.display[0]);
        render(
            <Provider store={reduxStore}>
                <Credential credential={credential} index={1} />
            </Provider>
        );
        const itemBoxElement = screen.getByTestId("ItemBox-Outer-Container");
        expect(itemBoxElement).toBeInTheDocument();
    });
    // test('check if content is rendered properly', () => {
    //     const clickHandler = jest.fn();
    //     const credential:CredentialWellknownObject = getCredentialObject();
    //     render(<Credential credential={credential} index={1} />);
    //     const itemBoxElement = screen.getByTestId("ItemBox-Outer-Container");
    //     expect(itemBoxElement).toHaveTextContent("TitleOfItemBox")
    // });
});
