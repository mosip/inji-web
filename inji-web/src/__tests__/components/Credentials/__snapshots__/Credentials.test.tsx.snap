// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the Layout of Credentials Check if the layout is matching with the snapshots 1`] = `
<DocumentFragment>
  <div
    class="bg-iw-tileBackground w-48 h-48 flex flex-col shadow hover:shadow-lg hover:scale-105 hover:shadow-iw-selectedShadow p-5 m-4 rounded-md cursor-pointer items-center"
    data-testid="ItemBox-Outer-Container-1"
    role="menuitem"
    tabindex="0"
  >
    <img
      alt="Issuer Logo"
      class="w-20 h-20 items-center justify-center me-4 flex"
      data-testid="ItemBox-Logo"
      src="https://url.com"
    />
    <div
      class="flex justify-center items-center pt-4"
    >
      <h3
        class="flex text-sm font-semibold text-iw-title text-center"
        data-testid="ItemBox-Text"
      >
        Name
      </h3>
    </div>
  </div>
</DocumentFragment>
`;
